services/src/handlers/upload.rs

```diff
@@ -0,0 +1,141 @@
+
+    let mut stream = MultipartStream::new(
+        boundary,
+        body.map_ok(|mut buf| buf.copy_to_bytes(buf.remaining())),
+    );
+
+    let upload_id = UploadId::new();
+
+    let root = upload_id.root_path()?;
+
+    fs::create_dir_all(&root).await.context(error::Io)?;
+
+    let mut files: Vec<FileUpload> = vec![];
+    while let Ok(Some(mut field)) = stream.try_next().await {
+        let file_name = field
+            .filename()
+            .map_err(|_| error::Error::UploadFieldMissingFileName)?
+            .to_owned();
+
+        let file_id = FileId::new();
+        let mut file = fs::File::create(root.join(&file_name))
+            .await
+            .context(error::Io)?;
+
+        let mut byte_size = 0;
+        while let Ok(Some(bytes)) = field.try_next().await {
+            file.write_all(&bytes).await.context(error::Io)?;
+            byte_size += bytes.len();
+        }
+
+        files.push(FileUpload {
+            id: file_id,
+            name: file_name,
+            byte_size,
+        });
+    }
+
+    ctx.data_set_db_ref_mut()
+        .await
+        .create_upload(
+            session.user.id,
+            Upload {
+                id: upload_id,
+                files,
+            },
+        )
+        .await?;
+
+    Ok(warp::reply::json(&IdResponse::from(upload_id)))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::contexts::InMemoryContext;
+    use crate::handlers::handle_rejection;
+    use crate::util::tests::create_session_helper;
+
+    #[tokio::test]
+    async fn upload() {
+        let ctx = InMemoryContext::default();
+
+        let session = create_session_helper(&ctx).await;
+
+        let body = r#"-----------------------------10196671711503402186283068890
+Content-Disposition: form-data; name="files[]"; filename="bar.txt"
+Content-Type: text/plain
+
+bar
+-----------------------------10196671711503402186283068890
+Content-Disposition: form-data; name="files[]"; filename="foo.txt"
+Content-Type: text/plain
+
+foo
+-----------------------------10196671711503402186283068890--
+"#
+        .to_string();
+
+        let res = warp::test::request()
+            .method("POST")
+            .path("/upload")
+            .header("Content-Length", body.len())
+            .header(
+                "Authorization",
+                format!("Bearer {}", session.id.to_string()),
+            )
+            .header("Content-Type", "multipart/form-data; boundary=---------------------------10196671711503402186283068890")
+            .body(
+                body,
+            )
+            .reply(&upload_handler(ctx).recover(handle_rejection))
+            .await;
+
+        assert_eq!(res.status(), 200);
+
+        let body = std::str::from_utf8(&res.body()).unwrap();
+        let _upload = serde_json::from_str::<IdResponse<UploadId>>(&body).unwrap();
+
+        // TODO: fix: body doesn't arrive at handler in test
```

Comment:
I don't use the warp-naitve multipart implementation, because that one is not stream-based, meaning all uploaded files are completely read into memory and only then written to disk. Instead I use mpart-async https://github.com/seanmonstar/warp/issues/323

In production the handler gets a stream but in the test I cannot give a stream as input. If I input the string directly, it doesn't arrive in the handler.

Reply:
But can't you create a stream then as input for the test?

Reply:
I did not succeed in doing so. The `.body` method the warp test framework provides only accepts `impl AsRef<[u8]>` inputs. Do you know a way to put a stream in there? 😉